File: thearcadefinalfinal/games/tictactoe/js/console-ui.js
/* ==========================================================================
   console-ui.js  —  port of ConsoleUI.java
   Same box-drawing board, same "B2"-style coordinate parser (column letter
   first, row digit second), same grumpy rejection message.
   ========================================================================== */
(function () {
  "use strict";
  var Player = window.T3.Player;

  function ConsoleUI(model, onMove) {
    this.model = model;
    this.onMove = onMove;
    this.out = null;
    this.input = null;
    this.log = [];
  }

  ConsoleUI.prototype.mount = function (container) {
    container.innerHTML = "";

    var term = document.createElement("div");
    term.className = "t3-terminal";

    var out = document.createElement("pre");
    out.className = "t3-term-out";
    out.setAttribute("aria-live", "polite");

    var row = document.createElement("div");
    row.className = "t3-term-row";

    var caret = document.createElement("span");
    caret.className = "t3-term-caret";
    caret.textContent = "ENTER MOVE >";

    var input = document.createElement("input");
    input.className = "t3-term-input";
    input.type = "text";
    input.maxLength = 2;
    input.autocomplete = "off";
    input.spellcheck = false;
    input.placeholder = "B2";
    input.setAttribute("aria-label", "Enter a move, for example B2");

    var self = this;
    input.addEventListener("keydown", function (e) {
      if (e.key !== "Enter") return;
      var s = input.value.trim().toUpperCase();
      input.value = "";
      if (!s) return;

      self.log.push("> " + s);
      if (!self.tryMoveFromString(s)) {
        self.log.push("Illegal Move. Give it another go bub");
        self.refresh();
      } else {
        self.onMove();
      }
    });

    row.appendChild(caret);
    row.appendChild(input);
    term.appendChild(out);
    term.appendChild(row);
    container.appendChild(term);

    this.out = out;
    this.input = input;
    this.refresh();
    input.focus();
  };

  ConsoleUI.prototype.tryMoveFromString = function (s) {
    if (s.length !== 2) return false;

    var columnChar = s.charAt(0);
    var rowChar = s.charAt(1);

    var column = { A: 0, B: 1, C: 2 }[columnChar];
    var row = { "1": 0, "2": 1, "3": 2 }[rowChar];

    if (column === undefined || row === undefined) return false;
    return this.model.move(row, column);
  };

  /* printBoard() */
  ConsoleUI.prototype.printBoard = function () {
    var lines = [];
    lines.push("    A   B   C");
    lines.push("  ┌───┬───┬───┐");
    for (var i = 0; i < 3; i++) {
      var line = (i + 1) + " |";
      for (var j = 0; j < 3; j++) {
        var p = this.model.getCell(i, j);
        var cell = (p === Player.NONE) ? " " : p;
        line += " " + cell + " |";
      }
      lines.push(line);
      if (i < 2) lines.push("  ├───┼───┼───┤");
    }
    lines.push("  └───┴───┴───┘");
    return lines.join("\n");
  };

  ConsoleUI.prototype.refresh = function () {
    if (!this.out) return;

    var text = this.printBoard() + "\n\n";

    if (this.model.isGameOver()) {
      text += (this.model.getWinner() !== Player.NONE)
        ? "Winner: " + this.model.getWinner() + "\n"
        : "Tie game!!\n";
      if (this.input) this.input.disabled = true;
    } else {
      text += "Turn for: " + this.model.getCurrentPlayer() + "\n";
      if (this.input) this.input.disabled = false;
    }

    // Keep the last few lines of chatter under the board, like scrollback.
    var tail = this.log.slice(-4);
    if (tail.length) text += "\n" + tail.join("\n");

    this.out.textContent = text;
  };

  ConsoleUI.prototype.clearLog = function () {
    this.log = [];
  };

  window.T3.ConsoleUI = ConsoleUI;
})();

File: java-source/tictactoe/ConsoleUI.java
import java.util.Locale;
import java.util.Scanner;

public class ConsoleUI {
    private final IT3Model model;
    private final Scanner in = new Scanner(System.in);
    public ConsoleUI(IT3Model model) {
        this.model = model;
        gameLoop();
    }

    private void gameLoop(){
        while (true){
            model.reset();
            while (!model.isGameOver()) {
                printBoard();
                System.out.println();
                System.out.println("Turn for: " + model.getCurrentPlayer());
                System.out.println("Enter move: ");
                String s = in.nextLine().trim().toUpperCase(Locale.ROOT);
                if (!tryMoveFromString(s)) {
                    System.out.println("Illegal Move. Give it another go bub");
                }
            }
            printBoard();
            if (model.getWinner() != Player.NONE){
                System.out.println("\nWinner: " + model.getWinner());
            } else {
                System.out.println("Tie game!!");
            }
            System.out.print("Play again? (y/n): ");
            String playAgain = in.nextLine().trim().toLowerCase(Locale.ROOT);
            if (!playAgain.equals("y")){
                break;
            }
        }
    }
    private boolean tryMoveFromString(String s){
        if (s.length() != 2){
            return false;
        }
        char columnChar = s.charAt(0);
        char rowChar = s.charAt(1);

        int column = switch (columnChar){
            case 'A' -> 0;
            case 'B' -> 1;
            case 'C' -> 2;
            default -> -1;
        };
        int row = switch (rowChar){
            case '1' -> 0;
            case '2' -> 1;
            case '3' -> 2;
            default -> -1;
        };
        if (row == -1 || column == -1){
            return false;
        }
        return model.move(row, column);
    }

    private void printBoard(){
        System.out.println("    A   B   C");
        System.out.println("  ┌───┬───┬───┐");
        for (int i = 0; i < 3; i++){
            System.out.print((i + 1) + " |");
            for (int j = 0; j < 3; j++){
                Player p = model.getCell(i,j);
                String cell = " ";
                if (p != Player.NONE){
                    cell = p.name();
                }
                System.out.print(" " + cell + " |");
            }
            System.out.println();
            if(i < 2){
                System.out.println("  ├───┼───┼───┤");
            }
        }
        System.out.println("  └───┴───┴───┘");
    }
}
